storage/src/index/storage.rs

```diff
@@ -1,337 +1,465 @@
-        }
+    /// Provides mutable access to the values associated with a key via a `Cursor`.
+    pub fn get_mut(&mut self, key: &[u8]) -> Option<Cursor<V>> {
+        let k = self.translator.transform(key);
+        self.map
+            .get_mut(&k)
+            .map(|record| Cursor::new(record, &self.collisions, &self.pruned))
     }
 
-    /// Retrieve all values associated with a translated key, allowing for mutation & insertion if
-    /// the key you are trying to update isn't currently active.
-    pub fn update_iter(&mut self, key: &[u8]) -> UpdateValueIterator<T::Key, V> {
-        let translated_key = self.translator.transform(key);
-        let entry = self.map.entry(translated_key);
-        match entry {
-            Entry::Occupied(mut occupied_entry) => {
-                let record_ptr = occupied_entry.get_mut();
-                UpdateValueIterator {
-                    next: Some(record_ptr),
-                    entry: Some(MapEntry::Occupied(occupied_entry)),
-                    collisions_counter: &self.collisions,
-                }
+    /// Remove all values at the given translated key.
+    pub fn remove(&mut self, key: &[u8]) {
+        // We use `prune()` to ensure we count all dropped values.
+        self.prune(key, |_| true);
+    }
+
+    /// Insert a value at the given translated key.
+    pub fn insert(&mut self, key: &[u8], v: V) {
+        let k = self.translator.transform(key);
+        match self.map.entry(k) {
+            Entry::Occupied(mut entry) => {
+                let record = entry.get_mut();
+                let mut cursor = Cursor::new(record, &self.collisions, &self.pruned);
+                cursor.next();
+                cursor.insert(v);
+            }
+            Entry::Vacant(entry) => {
+                entry.insert(Record {
+                    value: v,
+                    next: None,
+                });
             }
-            Entry::Vacant(vacant_entry) => UpdateValueIterator {
-                next: None,
-                entry: Some(MapEntry::Vacant(vacant_entry)),
-                collisions_counter: &self.collisions,
-            },
         }
     }
 
-    /// Retrieve all values associated with a translated key, allowing for mutation and removal.
-    pub fn remove_iter(&mut self, key: &[u8]) -> RemoveValueIterator<T::Key, V> {
-        let translated_key = self.translator.transform(key);
-        let entry = self.map.entry(translated_key);
-        match entry {
-            Entry::Occupied(mut occupied_entry) => {
-                let record_ptr = occupied_entry.get_mut();
-                RemoveValueIterator {
-                    can_remove: false,
-                    prev_prev: None,
-                    prev: None,
-                    next: Some(record_ptr),
-                    entry: Some(occupied_entry),
-                    pruned_counter: &self.keys_pruned,
+    /// Insert a value at the given translated key, and prune any values that are no longer valid.
```

Comment:
Might be worth clarifying: If the value you are inserting matches the predicate, will it also be pruned?